scalafmt-core/native/src/main/scala/org/scalafmt/internal/RegexCompat.scala

```diff
@@ -0,0 +1,222 @@
+        sb.append(']')
+        isInClass = false
+      case other =>
+        if (isEscaped) {
+          isEscaped = false
+          sb.append('\\')
+        }
+        sb.append(other)
+    }
+    sb.toString()
+  }
+
+  @inline
+  val trailingSpace = Pattern
+    .compile(fixHorizontalSpaceInRegex("\\h+$"), Pattern.MULTILINE)
+
+  // "slc" stands for single-line comment
+  @inline
+  val slcLine = Pattern
+    .compile(fixHorizontalSpaceInRegex("^/\\/\\/*\\h*(.*?)\\h*$"))
+
+  @inline
+  val slcDelim = Pattern.compile(fixHorizontalSpaceInRegex("\\h+"))
+
+  // "mlc" stands for multi-line comment
+  @inline
+  val mlcHeader = Pattern
+    .compile(fixHorizontalSpaceInRegex("^/\\*\\h*(?:\n\\h*[*]*\\h*)?"))
+
+  @inline
+  val mlcLineDelim = Pattern
+    .compile(fixHorizontalSpaceInRegex("\\h*\n\\h*[*]*\\h*"))
+
+  @inline
+  val mlcParagraphEnd = Pattern.compile("[.:!?=]$")
+
+  @inline
+  val mlcParagraphBeg = Pattern.compile("^(?:[-*@=]|\\d+[.:])")
+
+  @inline
+  val leadingAsteriskSpace = Pattern
+    .compile(fixHorizontalSpaceInRegex("\n\\h*[*][^*]"), Pattern.MULTILINE)
+
+  @inline
+  val docstringLine = Pattern.compile(
+    fixHorizontalSpaceInRegex("^(?:\\h*\\*)?(\\h*)(.*?)\\h*$"),
+    Pattern.MULTILINE,
+  )
+
+  @inline
+  val emptyLines = fixHorizontalSpaceInRegex("\\h*(\n\\h*\\*?\\h*)*")
+
+  @inline
+  val emptyDocstring = Pattern.compile(s"^/\\*\\*$emptyLines\\*/$$")
+
+  @inline
+  val onelineDocstring = {
+    val oneline = fixHorizontalSpaceInRegex("[^*\n\\h](?:[^\n]*[^\n\\h])?")
+    Pattern.compile(s"^/\\*\\*$emptyLines($oneline)$emptyLines\\*/$$")
+  }
+
+  @inline
+  val docstringLeadingSpace = Pattern.compile(fixHorizontalSpaceInRegex("^\\h+"))
+
+  @inline
+  def compileStripMarginPattern(pipe: Char) = Pattern
+    .compile(fixHorizontalSpaceInRegex(s"\n+\\h*?\\$pipe"), Pattern.MULTILINE)
+
+  @inline
+  def compileStripMarginPatternWithLineContent(pipe: Char) = Pattern.compile(
+    fixHorizontalSpaceInRegex(s"\n(\\h*\\$pipe)?([^\n]*)"),
+    Pattern.MULTILINE,
+  )
+
+  val stripMarginPatternWithLineContent =
+    compileStripMarginPatternWithLineContent('|')
+
+  // see: https://ammonite.io/#Save/LoadSession
+  @inline
+  val ammonitePattern: Regex = "(?:\\s*\\n@)".r
+
+  @inline
+  val stripMarginPattern = Pattern.compile(
+    fixHorizontalSpaceInRegex("\n(\\h*\\|)?([^\n]*)"),
+    Pattern.MULTILINE,
+  )
+
+  // startAfterPattern and endBeforePattern should be unique in basePattern
+  // basePattern = startAfterPattern + matched pattern + endBeforePattern
+  private def replaceAll(
+      basePattern: Pattern,
+      startAfterPattern: Pattern,
+      endBeforePattern: Pattern,
+      baseText: String,
+      replacingText: String,
+  ): String = {
+    val sb = new StringBuilder()
+    val matcher = basePattern.matcher(baseText)
+    var currPosition = 0
+    while (matcher.find()) {
+      val start = matcher.start()
+      val end = matcher.end()
+
+      sb.append(baseText.substring(currPosition, start))
```

Comment:
if you had used a java StringBuilder, you could have used the append method which takes offset and length so doesn't need to create a copy of the substring.